x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,421 @@
+    }
+
+    // These methods are protected for testing purposes:
+    protected void updateMapping(LongConsumer handler) {
+        Index leaderIndex = params.getLeaderShardId().getIndex();
+        Index followIndex = params.getFollowShardId().getIndex();
+
+        ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.clear();
+        clusterStateRequest.metaData(true);
+        clusterStateRequest.indices(leaderIndex.getName());
+
+        leaderClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(clusterStateResponse -> {
+            IndexMetaData indexMetaData = clusterStateResponse.getState().metaData().getIndexSafe(leaderIndex);
+            assert indexMetaData.getMappings().size() == 1;
+            MappingMetaData mappingMetaData = indexMetaData.getMappings().iterator().next().value;
+
+            PutMappingRequest putMappingRequest = new PutMappingRequest(followIndex.getName());
+            putMappingRequest.type(mappingMetaData.type());
+            putMappingRequest.source(mappingMetaData.source().string(), XContentType.JSON);
+            followerClient.admin().indices().putMapping(putMappingRequest, ActionListener.wrap(
+                putMappingResponse -> handler.accept(indexMetaData.getVersion()),
+                e -> handleFailure(e, () -> updateMapping(handler))));
+        }, e -> handleFailure(e, () -> updateMapping(handler))));
+    }
+
+    protected void innerSendBulkShardOperationsRequest(Translog.Operation[] operations,
+                                                     LongConsumer handler,
+                                                     Consumer<Exception> errorHandler) {
+        final BulkShardOperationsRequest request = new BulkShardOperationsRequest(params.getFollowShardId(), operations);
+        followerClient.execute(BulkShardOperationsAction.INSTANCE, request,
+            new ActionListener<BulkShardOperationsResponse>() {
+                @Override
+                public void onResponse(BulkShardOperationsResponse response) {
+                    handler.accept(response.getLocalCheckpoint());
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    errorHandler.accept(e);
+                }
+            }
+        );
+    }
+
+    protected void innerSendShardChangesRequest(long from,
+                                      Long to,
+                                      Consumer<ShardChangesAction.Response> handler,
+                                      Consumer<Exception> errorHandler) {
+        ShardChangesAction.Request request = new ShardChangesAction.Request(params.getLeaderShardId());
+        request.setMinSeqNo(from);
+        request.setMaxSeqNo(to);
+        request.setMaxTranslogsBytes(params.getMaxTranslogBytes());
+        leaderClient.execute(ShardChangesAction.INSTANCE, request, new ActionListener<ShardChangesAction.Response>() {
```

Comment:
you can use `ActionListener.wrap(handler::accept, errorHandler::accept)` instead

Reply:
I will do that